autowarefoundation/autoware.universe
# feat: publisher and server for calibration with external application

## Description


Add the following items for calibration with external applications.
- A dedicated publisher for calibration status
- A server to send a graph and new parameters for calibration.
- A client for server operational test.

## Related links


Need new .msg and .srv in [this PR](https://github.com/tier4/tier4_autoware_msgs/pull/19)
ADAPI adaptor [Link](https://github.com/tier4/tier4_ad_api_adaptor/pull/43)
Launcher for external ad_api_adaptor [Link](https://github.com/tier4/autoware_launch/pull/332)

## Tests performed


Preparation 
1. Make log.csv, accel_map.csv and brake_map.csv in the directory "install/accel_brake_map_calibrator/share/accel_brake_map_calibrator/config".


Test for calibration status

1. Change `update_suggest_thresh` in the below file from 0.7 to 0.9.(src/autoware/universe/vehicle/accel_brake_map_calibrator/accel_brake_map_calibrator/config/accel_brake_map_calibrator.param.yaml)
2. Run planning simulator.
3. Set initial position and goal.
4. Open the other terminal and run `ros2 run  accel_brake_map_calibrator sim_actuation_status_publisher.py`.
5. Open the other terminal and run `ros2 launch accel_brake_map_calibrator accel_brake_map_calibrator.launch.xml`.
6. Open the other terminal and run `ros2 topic echo /accel_brake_map_calibrator/output/calibration_status`.
7. Drive vehicle in the simulator for 500 meters or more (to collect driving data).
 
Calibration status will change from 0 (=NORMAL) to 1 (=WARNING) after the driving.

Test for the server
 Open two terminal and run the following.
 - Terminal for server : `ros2 launch accel_brake_map_calibrator accel_brake_map_calibrator.launch.xml`
 
 - Terminal for client : `ros2 run accel_brake_map_calibrator plot_csv_server_node.py`
     "plot.svg" should be created in the directory you run the script.
   The directory "test_data_save" should be created. It include  "accel_map.csv"", "brake_map.csv" and "graph.svg".

## Notes for reviewers



## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.


## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.


## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.


After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/
